### DISCUSSION TITLE
How do I handle Vercel preview and Supabase auth?

### DISCUSSION THREAD

**Discussion by pedrobslisboa** [supabase/supabase#2760] [Questions] [answered]
Hey guys, I've been looking for the answer to this question without success.

### How do I handle Vercel preview and Supabase auth?
E.g.: How can I have the redirect URL for auth with google provider and sync to Vercel?

<img width="1452" alt="Captura de Tela 2021-08-07 às 09 09 29" src="https://user-images.githubusercontent.com/35539594/128599666-849c104c-09ee-40bd-9a89-0d8d4ea0e454.png">

**Comment by silentworks** [comment]
You will have to add these manually or you can set the redirectTo dynamically inside of the Supabase Client API code.

For sign up
```js
const { error, data } = await supabase.auth.signUp({ provider: 'google' }, { redirectTo: '' });
```

For sign in
```js
const { error, data } = await supabase.auth.signIn({ provider: 'google' }, { redirectTo: '' });
```

**Reply by wookiehangover** [reply]
Setting the redirect dynamically as you suggested seems to only work if you also add the preview URL manually to the "Additional Redirect URLs" setting.

**Reply by lucasnantonio** [reply]
Thanks @silentworks for the help, but I don't get how this works. Bear with me and please let me know if I'm missing something:

1. I'm working on branch "new-feature" and create a new deploy on vercel. 
2. Vercel gives me a URL for the deploy. 
3. I add this URL manually to the Supabase "other domains" list.

Now...

Do I need to add this URL to the redirectTo call on supabaseclient? Because if I do so, it creates a new commit, which changes the deploy domain again on Vercel. 

Or is the URL parameter supposed to be literally ""? If so, I don't understand how supabase is supposed to know that the link should go to the new domain instead of the default domain.

**Reply by lucasnantonio** [reply]
I figured one way: created a constant URL like `preview.my-app.com` and added that to the redirect call and additional domains in Supabase. Then I can deploy form vercel always to that domain for testing.

**Comment by kennethcassel** [comment]
It would be great if we could set patterned URLS in the redirect settings. So then we could do something like `*.truthy.so` and pattern match allow all preview domains from our vercel deploys.

**Reply by activenode** [reply]
Yes, god please.

**Comment by activenode** [comment]
This here is a HUGE topic. Like HUGE. Why is this bumped so low? I mean all of continuous deployment is about having environments with different deploy URLs. 

This is literally _freaking_ me out right now as this is the last bit I'd be thinking of becoming a problem. I don't even get why there is a "baseURL". IMO from an architectural perspective the baseURL should be required in the init process of the supabase provider.

Then it will be constrained by the list of allowed wildcards within supabase config.

**Reply by multiplehats** [reply]
Agreed!

**Comment by MildTomato** [comment]
This issue is also being tracked [here](https://github.com/supabase/supabase/issues/4808) now

**Comment by kevcodez** [comment]
Supabase now supports wildcard URLs (see https://github.com/supabase/gotrue/pull/334). You can just use the glob pattern to define your URIs.

**Comment by thorwebdev** [comment]
sorry to hear you're having difficulty!

- `SITE_URL` = make this your official site URL
- Additional redirect URLs = should be something like this: 
  - `http://localhost:3000/*/*`
  - `https://**vercel.app/*/*` <- match Vercel deploy preview URLS

Then when you do a signIn you can do:

```js
const { data, error } = await supabase.auth.signInWithOAuth({
  provider: 'github'
  options: {
    redirectTo: `${
      process.env.NEXT_PUBLIC_VERCEL_URL
        ? 'https://' + process.env.NEXT_PUBLIC_VERCEL_URL
        : 'http://localhost:3000'
    }`
  }
}
```

`NEXT_PUBLIC_VERCEL_URL`: The URL of the deployment. Example: `my-site-7q03y4pi5.vercel.app`  -- https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables

**Comment by thorwebdev** [comment, marked as answer]
We've added a section in the docs for this: https://supabase.com/docs/guides/auth#redirect-urls-and-wildcards

**Reply by strangecarr** [reply]
hi, I'm using this code from the docs:

```
const getURL = () => {
  let url =
    process?.env?.NEXT_PUBLIC_SITE_URL ?? // Set this to your site URL in production env.
    process?.env?.NEXT_PUBLIC_VERCEL_URL ?? // Automatically set by Vercel.
    'http://localhost:3000/';
  // Make sure to include `https://` when not localhost.
  url = url.includes('http') ? url : `https://${url}`;
  // Make sure to including trailing `/`.
  url = url.charAt(url.length - 1) === '/' ? url : `${url}/`;
  return url;
};
```
as i set SITE_URL as an environment variable in Vercel, it looks like this code always picks up SITE_URL for previews so when on a preview URL, as soon as I auth I am then redirected to the production environment. does the code need to be modified to check the environment (e.g. deployment, preview)? Not an issue for local development.

**Reply by thorwebdev** [reply]
@strangecarr you need to make sure to not set `NEXT_PUBLIC_SITE_URL` in `preview` environment. It's a configuration option in your Vercel project settings. That way your preview branch picks up the `NEXT_PUBLIC_VERCEL_URL`.

<img width="889" alt="image" src="https://user-images.githubusercontent.com/5748289/230007867-24f2bf23-7a8a-4d69-a7b6-e45900ecf130.png">

**Reply by strangecarr** [reply]
Got it! Thanks so much for your quick reply. My bad, that saved me a lot of time and a messy hack!

**Reply by Justinfan827** [reply]
@thorwebdev how do you handle NEXT_PUBLIC_VERCEL_BRANCH_URL when you're working with a git branch URL? Both this env var and NEXT_PUBLIC_VERCEL_URL are set

**Reply by acomms** [reply]
Updated docs URL for those in 2024: https://supabase.com/docs/guides/auth/redirect-urls
Edit:
@Justinfan827 You can just add it to the nullish coalescing (??) check in the code provided. This is what I am doing:
```
function getRootURL() {
  let url =
    process?.env?.NEXT_PUBLIC_ROOT_DOMAIN_URL ?? // Set this to your site URL in production env.
    process?.env?.NEXT_PUBLIC_VERCEL_BRANCH_URL ?? // Automatically set by Vercel.
    process?.env?.NEXT_PUBLIC_VERCEL_URL ?? // Automatically set by Vercel.
    "http://localhost:3000";
  // Make sure to include `https://` when not localhost.
  url = url.includes("http") ? url : `https://${url}`;
  // Make sure to include a trailing `/`.
  url = url.charAt(url.length - 1) === "/" ? url : `${url}/`;
  return url;
}
```

**Reply by Justinfan827** [reply]
But both these env vars are set for both environments, so if you're on the `NEXT_PUBLIC_VERCEL_URL` env, this function will always route to `NEXT_PUBLIC_VERCEL_BRANCH_URL`

**Comment by nuotsu** [comment]
Does setup differ when using SvelteKit? Seems like this is for use with Next.js

**Reply by silentworks** [reply]
No the setup shouldn't differ.

**Comment by budivoogt** [comment]
I'm building a Svelte + Supabase todolist application. I've followed the instructions for defining the URL as detailed [here](https://supabase.com/docs/guides/auth#redirect-urls-and-wildcards). However my public magic links still aren't working.

Here is my getURL() snippet, modified to play nice with Svelte's Vite builder: https://gist.github.com/brucey0x/295a63fbb32b4346306535af93aab301

On the public production version hosted by Vercel, I can successfully see the `VITE_PUBLIC_SITE_URL` env which I've defined in Vercel and restricted to only be displayed in production. However the magic link keeps redirecting me to my `http://localhost:5173/` local host. Any suggestions on how to resolve?

Thank you!

**Reply by thibistaken** [reply]
Did you find a fix? Quite interested in this. Thanks!

**Reply by budivoogt** [reply]
I can't remember if I fixed it. Here's the [code in my repo](https://github.com/brucey0x/JQQ_todolist_app/blob/43c44a20ba6e5fb02e5d370cd99deba171f50da4/src/components/Auth.svelte) if you want to look around.
